src/extension/infer.rs

```diff
@@ -1133,6 +1160,502 @@ mod test {
+            root,
+            NodeType::open_extensions(ops::BasicBlock::Exit {
+                cfg_outputs: type_row![NAT],
+            }),
+        )?;
+
+        let entry = {
+            let dfb: OpType = ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: entry_ext.clone(),
+            }
+            .into();
+
+            let entry = hugr.add_node_before(exit, NodeType::open_extensions(dfb))?;
+            let entry_in = hugr.add_node_with_parent(
+                entry,
+                NodeType::open_extensions(ops::Input {
+                    types: type_row![NAT],
+                }),
+            )?;
+            let entry_out = hugr.add_node_with_parent(
+                entry,
+                NodeType::open_extensions(ops::Output {
+                    types: oneway(NAT).into(),
+                }),
+            )?;
+
+            let dfg = hugr.add_node_with_parent(
+                entry,
+                NodeType::open_extensions(make_opaque(
+                    PRELUDE_ID,
+                    FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&entry_ext),
+                )),
+            )?;
+
+            hugr.connect(entry_in, 0, dfg, 0)?;
+            hugr.connect(dfg, 0, entry_out, 0)?;
+            entry
+        };
+
+        let bb1 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: bb1_ext.clone(),
+            },
+            FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&bb1_ext),
+            make_opaque(
+                PRELUDE_ID,
+                FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&bb1_ext),
+            ),
+        )?;
+
+        let bb2 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: bb2_ext.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&bb2_ext),
+            make_opaque(
+                PRELUDE_ID,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&bb2_ext),
+            ),
+        )?;
+
+        hugr.connect(entry, 0, bb1, 0)?;
+        hugr.connect(bb1, 0, bb2, 0)?;
+        hugr.connect(bb1, 0, exit, 0)?;
+        hugr.connect(bb2, 0, entry, 0)?;
+
+        Ok(hugr)
+    }
+
+    #[test]
+    fn test_cfg_loops() -> Result<(), Box<dyn Error>> {
+        let just_a = ExtensionSet::singleton(&A);
+        let variants = vec![
+            (
```

Comment:
You could add a couple more here - with `just_a` repeated, and one with two blocks that do something. I mean, since you've set up this nifty framework already, I mean :-)